vault/identity_store_util.go

```diff
@@ -2120,3 +2191,196 @@ func (i *IdentityStore) memDBGroupsByBucketEntryKeyHashInTxn(txn *memdb.Txn, has
 
 	return groups, nil
 }
+
+func (i *IdentityStore) MemDBGroupByAliasIDInTxn(txn *memdb.Txn, aliasID string, clone bool) (*identity.Group, error) {
+	if aliasID == "" {
+		return nil, fmt.Errorf("missing alias ID")
+	}
+
+	if txn == nil {
+		return nil, fmt.Errorf("txn is nil")
+	}
+
+	alias, err := i.MemDBAliasByIDInTxn(txn, aliasID, false, true)
+	if err != nil {
+		return nil, err
+	}
+
+	if alias == nil {
+		return nil, nil
+	}
+
+	return i.MemDBGroupByIDInTxn(txn, alias.ParentID, clone)
+}
+
+func (i *IdentityStore) MemDBGroupByAliasID(aliasID string, clone bool) (*identity.Group, error) {
+	if aliasID == "" {
+		return nil, fmt.Errorf("missing alias ID")
+	}
+
+	txn := i.db.Txn(false)
+
+	return i.MemDBGroupByAliasIDInTxn(txn, aliasID, clone)
+}
+
+func (i *IdentityStore) deleteGroupAlias(aliasID string) error {
+	var err error
+	var alias *identity.Alias
+	var group *identity.Group
+
+	if aliasID == "" {
+		return fmt.Errorf("missing alias ID")
+	}
+
+	i.groupLock.Lock()
+	defer i.groupLock.Unlock()
+
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	alias, err = i.MemDBAliasByIDInTxn(txn, aliasID, false, true)
+	if err != nil {
+		return err
+	}
+
+	if alias == nil {
+		return nil
+	}
+
+	group, err = i.MemDBGroupByAliasIDInTxn(txn, alias.ID, true)
+	if err != nil {
+		return err
+	}
+
+	// If there is no group tied to a valid alias, something is wrong
+	if group == nil {
+		return fmt.Errorf("alias not associated to a group")
+	}
+
+	// Delete group alias in memdb
+	err = i.MemDBDeleteAliasByIDInTxn(txn, group.Alias.ID, true)
+	if err != nil {
+		return err
+	}
+
+	// Delete the alias
+	group.Alias = nil
+
+	err = i.upsertGroupInTxn(txn, group, true)
+	if err != nil {
+		return err
+	}
+
+	txn.Commit()
+
+	return nil
+}
+
+func (i *IdentityStore) refreshExternalGroupMembershipsByEntityID(entityID string, groupAliases []*logical.Alias) error {
+	if entityID == "" {
+		return fmt.Errorf("empty entity ID")
+	}
+
+	i.groupLock.Lock()
+	defer i.groupLock.Unlock()
+
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	oldGroups, err := i.MemDBGroupsByMemberEntityIDInTxn(txn, entityID, true, true)
+	if err != nil {
+		return err
+	}
+
+	var newGroups []*identity.Group
+	for _, alias := range groupAliases {
+		aliasByFactors, err := i.MemDBAliasByFactors(alias.MountAccessor, alias.Name, true, true)
+		if err != nil {
+			return err
+		}
+		if aliasByFactors == nil {
+			continue
+		}
+		mappingGroup, err := i.MemDBGroupByAliasID(aliasByFactors.ID, true)
+		if err != nil {
+			return err
+		}
+		if mappingGroup == nil {
+			return fmt.Errorf("group unavailable for a valid alias ID %q", aliasByFactors.ID)
+		}
+		newGroups = append(newGroups, mappingGroup)
```

Comment:
This doesn't need to be an append, can be a straight assignment and save the allocation.

Reply:
I am sorry. Not sure if I understand you here. `mappingGroup` is a group object to be added to the `newGroups` slice. We can do it without using append?

Reply:
Oops! NM. I misread the `for` as an `if`